device/Microsoft.Azure.Devices.Client/Transport/RetryDelegatingHandler.cs

```diff
@@ -12,14 +12,15 @@ namespace Microsoft.Azure.Devices.Client.Transport
     using Microsoft.Azure.Devices.Client.Exceptions;
     using Microsoft.Azure.Devices.Client.TransientFaultHandling;
     using Microsoft.Azure.Devices.Shared;
+    using System.Diagnostics;
 
     internal class RetryDelegatingHandler : DefaultDelegatingHandler
     {
         const int UndeterminedPosition = -1;
         const string StopRetrying = "StopRetrying";
 
         // RetryCount is used for testing purpose and is equal to MaxValue in prod.
-        internal static int RetryCount = Int32.MaxValue;
+        internal static int RetryCount = 10;
```

Comment:
>10 [](start = 41, length = 2)

This is breaking change for current customer who is doing infinite retries.  The existing behavior is that retry is limited by operation timeout ONLY...but with this change, retry is also limited by the retry count which means that retry could stop before the operation timeout...  This is in violation of spec.

Reply:
@yzhong94 @tameraw @olivierbloch In that case I think the spec is wrong. For more details and my recommendation for IoT developers in case of permanent network failures, please see https://github.com/Azure/azure-iot-sdk-csharp/issues/85#issuecomment-350385365

Reply:
Just so I'm clear, MaxValue is set by the customer right?  If it's not set, then it will do infinite retries?  In this case, why should we hard code it to 10?

Reply:
A retry policy can be defined by the customer (see the recommended workaround in https://github.com/Azure/azure-iot-sdk-csharp/issues/85#issuecomment-350385365).

This is for the default retry count which should not be infinite (or MaxValue).

Reply:
Default retry count should be the number of retries fitting into the default 4 minutes window (or whatever specified by the user) using a specific retry strategy.  It shouldn't be hardcoded to 10 either...

Reply:
> Default retry count should be the number of retries fitting into the default 4 minutes window

Why should we mix number of retries with retry times? Right now, if the request fails immediately we execute for about 1 minute with most likely no chance of recovery (e.g. if the cable is pulled):

![image](https://user-images.githubusercontent.com/8635911/33967879-35d15c7c-e01a-11e7-8b73-1d0bd4d585ae.png)

Instead of doing this, we should allow the application to understand what the failure is and stop retrying due to the number of retries failed not the 4 minute timeout. (In this case it should get the SocketException with media failure instead of getting a TimeoutException.)

Reply:
I am only speaking from the view point of what the spec has defined which was previously agreed upon within the team.  I will leave it to @yzhong94 and  @olivierbloch to comment.  We will need to sync this with other SDK owners if we decided to make any modification ...

Reply:
Agree, if we are changing the behavior of something already released, we need to have a discussion.

Reply:
@jasmineymlo @yzhong94 I've discussed this offline with @olivierbloch and @jebrando and agreed to add a minimum backoffTime + executionTime in case the operation completes very fast.
The new default retry curve for the case where the operation completes immediately is:

![image](https://user-images.githubusercontent.com/8635911/33972669-ebe2c366-e033-11e7-9a4b-c1f993fb8a0c.png)

I've added back the infinite retry and opened #290 to track proper error reporting after retry.

Reply:
Okay, thanks!  I like this solution